fastcrypto/src/secp256r1.rs

```diff
@@ -0,0 +1,566 @@
+        let k = Scalar::from(ScalarCore::<NistP256>::new(*k).unwrap());
+
+        if k.borrow().is_zero().into() {
+            return Err(signature::Error::new());
+        }
+
+        let z = Scalar::from_be_bytes_reduced(z);
+
+        // Compute scalar inversion of 𝑘
+        let k_inv = Option::<Scalar>::from(k.invert()).ok_or_else(signature::Error::new)?;
+
+        // Compute 𝑹 = 𝑘×𝑮
+        let big_r = (ProjectivePoint::GENERATOR * k.borrow()).to_affine();
+
+        // Lift x-coordinate of 𝑹 (element of base field) into a serialized big
+        // integer, then reduce it into an element of the scalar field
+        let r = Scalar::from_be_bytes_reduced(big_r.x());
+
+        let x = Scalar::from_be_bytes_reduced(x.to_be_byte_array());
+
+        // Compute 𝒔 as a signature over 𝒓 and 𝒛.
+        let s = k_inv * (z + (r * x));
+
+        if s.is_zero().into() {
+            return Err(signature::Error::new());
+        }
+
+        let sig = ExternalSignature::from_scalars(r, s)?;
+
+        // Note: This line is introduced here because big_r.y is a private field.
+        let y: Scalar = Scalar::from_be_bytes_reduced(*big_r.to_encoded_point(false).y().unwrap());
+
+        // Compute recovery id and normalize signature
+        let is_r_odd = y.is_odd();
+        let is_s_high = sig.s().is_high();
+        let is_y_odd = is_r_odd ^ is_s_high;
+        let sig_low = sig.normalize_s().unwrap_or(sig);
+        let recovery_id = RecoveryId::new(is_y_odd.into(), false);
+
+        Ok(Secp256r1Signature {
+            sig: sig_low,
+            bytes: OnceCell::new(),
+            recovery_id: recovery_id.to_byte(),
+        })
+    }
+}
+
+impl TryFrom<Secp256r1PublicKeyBytes> for Secp256r1PublicKey {
+    type Error = signature::Error;
+
+    fn try_from(bytes: Secp256r1PublicKeyBytes) -> Result<Secp256r1PublicKey, Self::Error> {
+        Secp256r1PublicKey::from_bytes(bytes.as_ref()).map_err(|_| Self::Error::new())
+    }
+}
+
+impl From<&Secp256r1PublicKey> for Secp256r1PublicKeyBytes {
+    fn from(pk: &Secp256r1PublicKey) -> Self {
+        Secp256r1PublicKeyBytes::from_bytes(pk.as_ref()).unwrap()
+    }
+}
+
+impl From<Secp256r1PrivateKey> for Secp256r1KeyPair {
+    fn from(secret: Secp256r1PrivateKey) -> Self {
+        let name = Secp256r1PublicKey::from(&secret);
+        Secp256r1KeyPair { name, secret }
+    }
+}
+
+impl Secp256r1Signature {
+    /// Recover the public used to create this signature. This assumes the recovery id byte has been set.
+    ///
+    /// This is copied from `recover_verify_key_from_digest_bytes` in the k256@0.11.6 crate except for a few additions.
+    ///
+    /// An [FastCryptoError::GeneralError] is returned if no public keys can be recovered.
+    pub fn recover(&self, msg: &[u8]) -> Result<Secp256r1PublicKey, FastCryptoError> {
+        let (r, s) = self.sig.split_scalars();
+        let v = RecoveryId::from_byte(self.recovery_id).unwrap();
```

Comment:
As this will run on blockchain, let's ensure it returns an error vs panic on fail, unless we know it will never fail, even under malicious signers.

Reply:
That's true. The deserialization doesn't check the recovery id byte, so a malicious signer could just put anything there. It's fixed now.